vhdx/src/vhdx.rs

```diff
@@ -0,0 +1,228 @@
+    fn read(&mut self, buf: &mut [u8]) -> std::result::Result<usize, std::io::Error> {
+        let sector_count =
+            div_round_up!(buf.len() as u64, self.disk_spec.logical_sector_size as u64);
+        let sector_index = self.current_offset / self.disk_spec.logical_sector_size as u64;
+
+        match vhdx_io::read(
+            &mut self.file,
+            buf,
+            &self.disk_spec,
+            &self.bat_entries,
+            sector_index,
+            sector_count,
+        ) {
+            Ok(r) => Ok(r),
+            Err(e) => {
+                println!("VHDx read: {}", e);
+                Err(io::Error::last_os_error())
+            }
+        }
+    }
+}
+
+impl Write for Vhdx {
+    fn flush(&mut self) -> std::result::Result<(), std::io::Error> {
+        self.file.flush()
+    }
+
+    /// Wrapper function to satisfy the Write trait implementation for VHDx disk. Converts the offest to secotor index and buffer length to sector count.
+    fn write(&mut self, buf: &[u8]) -> std::result::Result<usize, std::io::Error> {
+        let sector_count =
+            div_round_up!(buf.len() as u64, self.disk_spec.logical_sector_size as u64);
+        let sector_index = self.current_offset / self.disk_spec.logical_sector_size as u64;
+
+        match vhdx_io::write(
+            &mut self.file,
+            buf,
+            &self.disk_spec,
+            self.bat_entry.file_offset,
+            &mut self.bat_entries,
+            sector_index,
+            sector_count,
+        ) {
+            Ok(r) => Ok(r),
+            Err(e) => {
+                println!("Vhdx write: {}", e);
+                Err(io::Error::last_os_error())
+            }
+        }
+    }
+}
+
+impl Seek for Vhdx {
+    /// Wrapper function to satisfy the Seek trait implementation for VHDx disk. Updates the offset field in the Vhdx struct.
+    fn seek(&mut self, pos: SeekFrom) -> std::io::Result<u64> {
+        let new_offset: Option<u64> = match pos {
+            SeekFrom::Start(off) => Some(off),
+            SeekFrom::End(off) => {
+                if off < 0 {
+                    0i64.checked_sub(off).and_then(|increment| {
+                        self.virtual_disk_size().checked_sub(increment as u64)
+                    })
+                } else {
+                    self.virtual_disk_size().checked_add(off as u64)
+                }
+            }
+            SeekFrom::Current(off) => {
+                if off < 0 {
+                    0i64.checked_sub(off)
+                        .and_then(|increment| self.current_offset.checked_sub(increment as u64))
+                } else {
+                    self.current_offset.checked_add(off as u64)
+                }
+            }
+        };
+
+        if let Some(o) = new_offset {
+            if o <= self.virtual_disk_size() {
+                self.current_offset = o;
+                return Ok(o);
+            }
+        }
+        Err(std::io::Error::from_raw_os_error(EINVAL))
```

Comment:
Use https://doc.rust-lang.org/std/io/enum.ErrorKind.html#variant.InvalidData instead.

Reply:
The Seek trait expects std::io::Error from seek(), where InvalidData is std::io::ErrorKind.

Reply:
Of course it expects `std::io::Error`, I was suggesting to create a new `std::io::Error` with the `ErrorKind` `InvalidData`.
```suggestion
        Err(std::io::Error::new(std::io::ErrorKind::InvalidData, format!("Failed seeking to {}", pos)))
```

Reply:
Thanks for the solution. Is it ok if we do not print pos? It is not an integer but a SeekFrom structure.